peekingduck/pipeline/nodes/heuristic/fps.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2021 AI Singapore
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Calculates the FPS of video
+"""
+
+from typing import Any, Dict, List
+from statistics import mean
+from time import perf_counter
+
+from peekingduck.pipeline.nodes.node import AbstractNode
+
+NUM_FRAMES = 10
+
+
+class Node(AbstractNode):
+    """ FPS node class that calculates the FPS of the image frame """
+
+    def __init__(self, config: Dict[str, Any]) -> None:
+        super().__init__(config, node_path=__name__)
+
+        self.fps_type = config["dampen_fps"]
+        self.fps_log_display = config["fps_log_display"]
+        self.average_fps_log_freq = config["fps_log_freq"]
+
+        self.count = 0
+        self.global_avg_fps = 0.0
+        self.prev_frame_timestamp = 0.0
+        self.moving_average_fps: List[float] = []
+
+        if self.fps_log_display:
+            self.logger.info('FPS Moving Average will be logged every: %s frames', \
+                             self.average_fps_log_freq)
+
+    def run(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+        """ Calculates FPS using the time difference between the current
+        frame and the previous frame.
+
+        Args:
+            inputs: ["pipeline_end"]
+
+        Returns:
+            outputs (dict): Dict with key "fps".
+        """
+
+        curr_frame_timestamp = perf_counter()
+
+        # Frame level FPS
+        frame_fps = 1./(curr_frame_timestamp - self.prev_frame_timestamp)
+        self.prev_frame_timestamp = curr_frame_timestamp
+
+        # Calculate global cumulative moving average
+        self.global_avg_fps = self._global_average(frame_fps)
+
+        # Calculate moving average FPS (dampen_fps)
+        average_fps = self._moving_average(frame_fps)
+
+        if self.fps_log_display:
+            if self.count%self.average_fps_log_freq == 0:
+                self.logger.info('10-frame Moving Average FPS: %.2f', average_fps)
```

Comment:
Moving average and global/cumulative avg are terms that we use internally, it might be puzzling to regular users. This seems to be clearer

![image](https://user-images.githubusercontent.com/33995028/126577681-3f231e5a-9331-4e64-9f9e-945a2aa76f43.png)

Reply:
Updated